src/govuk/components/checkboxes/checkboxes.js

```diff
@@ -29,13 +30,30 @@ Checkboxes.prototype.init = function () {
     // If we have content that is controlled, set attributes.
     $input.setAttribute('aria-controls', controls)
     $input.removeAttribute('data-aria-controls')
-    this.setAttributes($input)
-  }.bind(this))
+  })
+
+  if (document.readyState === 'complete') {
```

Comment:
Should this be 
```
document.addEventListener('readystatechange', this.syncState.bind(this))
```
```
Checkboxes.prototype.syncState = function () {
  if (document.readyState === 'complete') {
```
because at the moment it's only evaluated once (when it's `loading`) so `complete` is never hit. But I might be misunderstanding the reason for this check.

Reply:
The intention of this logic is to handle cases where the script is being evaluated after the `DOMContentLoaded` event has already fired (in which case `document.readyState` would be `complete`).

This might be unnecessary – I am not sure if the way the script is included means this is guaranteed to run before the `DOMContentLoaded`event fires. However, given this code is probably going to end up being included in ways we can't anticipate (e.g. as part of a React component) I'm thinking it might be safer to try and cater for these scenarios?

Having done a bit more reading, I think that I possibly need to flip this in order to handle readyState being interactive, like in [this example in the MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/Document/DOMContentLoaded_event#Checking_whether_loading_is_already_complete):

```js
if (document.readyState === 'loading') {
  window.addEventListener('DOMContentLoaded', this.syncState.bind(this))
} else {
  this.syncState()
}
```

Reply:
Thanks for digging into this. It seems that [older IE only support `complete` for `readyState`](https://caniuse.com/#search=readystate) so according to that `loading` wouldn't ever be true for those browsers.

Reply:
That page suggests that IE9-10 can fire the 'interactive' state too early, but also that they only support 'complete' – how can both of those be true?

It also associates the bug about Internet Explorer 9 and 10 with the IE11 block?

MDN seems to suggest that it's only IE8 that only supports `complete`: https://developer.mozilla.org/en-US/docs/Web/API/Document/readyState

Reply:
Raised an issue about the misleading notes – https://github.com/Fyrd/caniuse/issues/5489

Reply:
So because of the bug in IE9-10 the readyState as being reported as `interactive` too early, which means that the state is not being restored correctly in those browsers.

I don't think it's safe to just wait for the `DOMContentLoaded` because we can't guarantee it hasn't already fired – an example might be if a user was initialising the checkboxes on a page that had already loaded, e.g. if they were added dynamically.

We could consider syncing within `init`, on `DOMContentLoaded` _and_ on the `pageshow` event, but wondering if it's worth it given that we're only aiming for 'functional' support of IE9 and IE10 and this is something of an edge case?

Reply:
At the very least, I'll need to revise the commit message and PR description as this it currently implies that it fixes IE8-11, which is not the case.

Reply:
Actually, maybe the correct approach is to:

- always sync within the `init` function
- sync on the `pageshow` event OR the `DOMContentLoaded` if the browser does not support the `pageshow` event

Trying that now.

Reply:
This sounds like a good solution, 🙌 for the investigation and write up @36degrees.